src/graphics/program-lib/programs/standard-node.js

```diff
@@ -0,0 +1,93 @@
+import { standard } from './standard.js';
+
+// just do a shallow copy standard
+var standardnode = Object.assign({}, standard);
+
+// override a couple of functions - the overridden functions have been altered in standard to accept extra params
+standardnode._generateKey = standard.generateKey;
+standardnode.generateKey = function (options) {
+    return this._generateKey(options, options._shaderGraphChunk.id);
+};
+
+standardnode._createShaderDefinition = standard.createShaderDefinition;
+standardnode.createShaderDefinition = function (device, options) {
+    var rootShaderGraph = null;
+    var rootDeclGLSL = '';
+    var rootCallGLSL = '';
+
+    if (options._shaderGraphChunk) {
+        rootShaderGraph = options._shaderGraphChunk;
+
+        rootDeclGLSL = rootShaderGraph.generateRootDeclGlsl();
+        rootCallGLSL = rootShaderGraph.generateRootCallGlsl();
+    }
+
+    var graphCode = [];
+
+    if (options._shaderGraphChunk) {
+        graphCode[0] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_vertOff')) {
+            graphCode[0] += "#define SG_VS\n";
+            graphCode[0] += rootDeclGLSL;
+        }
+    }
+
+    if (options._shaderGraphChunk) {
+        graphCode[1] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_vertOff')) {
+            graphCode[1] += rootCallGLSL;
+            graphCode[1] += "   vPositionW = vPositionW+OUT_vertOff;\n";
+            graphCode[1] += "   gl_Position = matrix_viewProjection*vec4(vPositionW,1);\n";
+        }
+    }
+
+    if (options._shaderGraphChunk) {
+        graphCode[2] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_dAlpha') || rootShaderGraph.getIoPortByName('OUT_dNormalMap') || rootShaderGraph.getIoPortByName('OUT_dGlossiness') || rootShaderGraph.getIoPortByName('OUT_dSpecularity') || rootShaderGraph.getIoPortByName('OUT_dAlbedo') || rootShaderGraph.getIoPortByName('OUT_fragOut') || rootShaderGraph.getIoPortByName('OUT_dEmission')) {
+            graphCode[2] += "#define SG_PS\n";
+            graphCode[2] += rootDeclGLSL;
+        }
+    }
+
+    if (options._shaderGraphChunk) {
+        graphCode[3] = '';
+        if (rootShaderGraph.getIoPortByName('OUT_dAlpha') || rootShaderGraph.getIoPortByName('OUT_dNormalMap') || rootShaderGraph.getIoPortByName('OUT_dGlossiness') || rootShaderGraph.getIoPortByName('OUT_dSpecularity') || rootShaderGraph.getIoPortByName('OUT_dAlbedo') || rootShaderGraph.getIoPortByName('OUT_fragOut') || rootShaderGraph.getIoPortByName('OUT_dEmission')) {
```

Comment:
this function should be made more generic based on possible shader node outputs being specified in data, mentioned elsewhere

Reply:
I'll think about how to do this - however this code defines the chunks that gets inserted into the Standard Material's via its shader generator - so I wanted to keep it as simple as possible - with some of the same searchable terms in the standard material.

Reply:
I have improved this - I haven't turned it into a generic data driven system because it is still tied to the non-data driven standard material.